### ISSUE TITLE
[Impeller] Empty subpasses should not be elided when they have "destructive" modifiers

### ISSUE THREAD

**Issue by bdero** [flutter/flutter#126391]
Currently, if a SaveLayer contains no Entities, we always skip rendering it because it has no coverage. However, this is incorrect in cases where the entire subpass will be modified:

* The subpass blend mode is destructive.
* ~~The subpass has a destructive color filter.~~ (edit moved to issue https://github.com/flutter/flutter/issues/137090)
* The subpass has a clear color set on it.

For all of these cases, the subpass also needs to [expand to fill the current clip](https://github.com/flutter/flutter/issues/126389).

**Comment by bdero** [comment]
We also shouldn't skip if the EntityPass clear color is not fully transparent black.

**Comment by chinmaygarde** [comment]
We can file bugs with specific cases where we should not be skipping entities. Or perhaps this is an umbrella bug? Closing as not immediately actionable.

**Comment by bdero** [comment]
I made the title of the issue less ambiguous and updated the description to enumerate all of the cases that need to be handled. Let me know if this still seems too unclear to be actionable.

**Comment by gaaclarke** [comment]
My understanding from the issue is that the following should render a blue square on a red background:
```c++
TEST_P(AiksTest, EmptySaveLayerRenders) {
  Canvas canvas;
  canvas.Scale(GetContentScale());
  canvas.DrawPaint(Paint{.color = Color::Red()});
  canvas.ClipRect({100, 100, 200, 200});
  canvas.SaveLayer(Paint{.color = Color::Blue()});
  canvas.Restore();
  ASSERT_TRUE(OpenPlaygroundHere(canvas.EndRecordingAsPicture()));
}
```

**Comment by gaaclarke** [comment]
Tweaking to `EntityPass::GetSubpassCoverage` to return the `coverage_limit` makes the coverage of the subpass correct.  That is part of the solution since right now it return `nullopt` without any entities.  However the texture isn't rendered correctly too, so it is just a magenta texture.  So there needs to be a fix there as well.

<img width="781" alt="Screenshot 2023-10-17 at 10 39 39 AM" src="https://github.com/flutter/flutter/assets/30870216/ee7135b2-0528-4057-aac0-12060bd072ec">

**Comment by gaaclarke** [comment]
@bdero what do you think about this fix for the solid color case:
1) Change `EntityPass::GetSubpassCoverage` to return `coverage_limit` if the clear color is not transparent
1) Change `EntityPass::OnRender` to generate and render an entity with a SolidColorContents with the clear color if `elements_.empty()`

Sound good?

**Comment by gaaclarke** [comment]
If you following the `Paint::color` argument of the `SaveLayer`, it can be seen that that value is thrown away.  That has to be stored somewhere to be meaningful so that `EntityPass::GetClearColor` can use that as its default instead of defaulting to black-clear.

**Comment by gaaclarke** [comment]
@bdero I wrote up what I described in https://github.com/flutter/engine/pull/47026.  Please give that a check.  It renders correctly, I just want to make sure that's the approach we want.

**Comment by bdero** [comment]
Thanks! Took a look [here](https://github.com/flutter/engine/pull/47026). The coverage change makes sense but we shouldn't need to use a draw call to clear the texture.

**Comment by gaaclarke** [comment]
> The subpass has a clear color set on it.

We established that that isn't how skia works: https://fiddle.skia.org/c/3a9c6e7fcae654a1b8677fe9c3784b1f

> The subpass blend mode is destructive.

Here skia is inconsistent.  With a [kClear](https://fiddle.skia.org/c/a7b3a65bf125073b59061d6cc09599a8) blend mode it is rendered, with [kMultiply](https://fiddle.skia.org/c/3a9c6e7fcae654a1b8677fe9c3784b1f) it isn't.  We already appear to be matching skia there.

**Comment by gaaclarke** [comment]
> The subpass has a destructive color filter.

[This](https://fiddle.skia.org/c/6f18804281e6168b0203bb3ce7cf3474) appears to be ignored in skia too.  I don't think there is anything to do.  @bdero as long as that sounds good to you, I'll just add a PR that adds golden image tests for these cases to make sure we continue to work properly.

**Comment by bdero** [comment]
We happened to solve the destructive color case at some point after this bug was made.

One case we're definitely not handling yet is destructive color filters. For this I'd recommend adding a way to detect this on the Aiks `ColorFilter` abstraction. Something like:
* Move `Entity::IsBlendModeDestructive` to live in color.cc instead. We've already moved pretty much all the other blend stuff there.
* Add a virtual `IsDestructive()` to `ColorFilter`, `ColorFilter` impl just returns `false`.
* In the `ColorFilterMatrix` override, return true of the set color matrix would result in anything other than the destination color returning when the source color if fully transparent.
* For the `BlendColorFilter`, same deal if the set blend mode is destructive.
* Add an `bool IsColorFilterDestructive()` to the pass delegate that returns true if the paint's `ColorFilter` is destructive.
* Adjust our logic for expanding the subpass coverage [here](https://github.com/flutter/engine/blob/3b0469b0c718cdd69570c0830743460d5f159200/impeller/entity/entity_pass.cc#L613-L616) to account for destructive color filters here.
